prov/efa/src/rdm/efa_rdm_ep_progress.c

```diff
@@ -465,25 +466,39 @@ static inline void efa_rdm_ep_poll_ibv_cq(struct efa_rdm_ep *ep, size_t cqe_to_p
 	while (!err) {
 		pkt_entry = (void *)(uintptr_t)ep->ibv_cq_ex->wr_id;
 		efa_rdm_tracepoint(poll_cq, (size_t) ep->ibv_cq_ex->wr_id);
+		opcode = ibv_wc_read_opcode(ep->ibv_cq_ex);
 		if (ep->ibv_cq_ex->status) {
 			prov_errno = ibv_wc_read_vendor_err(ep->ibv_cq_ex);
-			opcode = ibv_wc_read_opcode(ep->ibv_cq_ex);
-			if (opcode == IBV_WC_SEND || opcode == IBV_WC_RDMA_WRITE) {
+			switch (opcode) {
+			case IBV_WC_SEND:
 #if ENABLE_DEBUG
-			if (opcode == IBV_WC_SEND)
 				ep->failed_send_comps++;
-			else
+#endif
+				efa_rdm_pke_handle_tx_error(pkt_entry, FI_EIO, prov_errno);
+				break;
+			case IBV_WC_RDMA_WRITE:
+#if ENABLE_DEBUG
 				ep->failed_write_comps++;
 #endif
 				efa_rdm_pke_handle_tx_error(pkt_entry, FI_EIO, prov_errno);
-			} else {
-				assert(opcode == IBV_WC_RECV);
+				break;
+			case IBV_WC_RDMA_READ:
+#if ENABLE_DEBUG
+				ep->failed_read_comps++;
+#endif
+				efa_rdm_pke_handle_tx_error(pkt_entry, FI_EIO, prov_errno);
+				break;
```

Comment:
nit we can also do fallthrough e.g.
```c
case IBV_WC_SEND:
#if ENABLE_DEBUG
    ep->failed_send_comps++;
#endif
case IBV_WC_RDMA_WRITE:
#if ENABLE_DEBUG
    ep->failed_write_comps++;
#endif
case IBV_WC_RDMA_READ:
#if ENABLE_DEBUG
    ep->failed_read_comps++;
#endif
    efa_rdm_pke_handle_tx_error(pkt_entry, FI_EIO, prov_errno);
    break;
```

Reply:
doesn't that will make `failed_write_comps` incremented for send as well?

Reply:
I see. The counter indeed introduces extra coding. We can find a more concise approach but I don't have a strong preference.

Reply:
yes, ideally we should declare `failed_comps` as an array, and have a map between the ibv wc code to the ofi op code.

Reply:
offline talked with @wenduwan, new revision removed the unused failed_send/write_comps to reduce the code duplication